Modules/StudyProgramme/classes/class.ilObjStudyProgramme.php

```diff
@@ -1402,6 +1430,170 @@ public function isAutoContentApplicable() : bool
 	}
 
 
+	////////////////////////////////////
+	// AUTOMATIC MEMBERSHIPS
+	////////////////////////////////////
+
+	/**
+	 * Get sources for auto-memberships.
+	 * @return ilStudyProgrammeAutoMembershipSource[]
+	 */
+	public function getAutomaticMembershipSources(): array
+	{
+		return $this->auto_memberships_repository->readFor($this->getId());
+	}
+
+	/**
+	 * Store a source to be monitored for automatic memberships.
+	 * @param string $type
+	 * @param int $src_id
+	 */
+	public function storeAutomaticMembershipSource(string $type, int $src_id) {
+		$ams = $this->auto_memberships_repository->create($this->getId(), $type, $src_id, false);
+		$this->auto_memberships_repository->update($ams);
+	}
+
+	/**
+	 * Delete a membership source.
+	 * @param string $type
+	 * @param int $src_id
+	 */
+	public function deleteAutomaticMembershipSource(string $type, int $src_id)
+	{
+		return $this->auto_memberships_repository->delete($this->getId(), $type, $src_id);
+	}
+
+	/**
+	 * Delete all membership sources of this StudyProgramme;
+	 */
+	public function deleteAllAutomaticMembershipSources()
+	{
+		return $this->auto_memberships_repository->deleteFor($this->getId());
+	}
+
+	/**
+	 * Disable a membership source.
+	 * @param string $type
+	 * @param int $src_id
+	 */
+	public function disableAutomaticMembershipSource(string $type, int $src_id)
+	{
+		$ams = $this->auto_memberships_repository->create($this->getId(), $type, $src_id, false);
+		$this->auto_memberships_repository->update($ams);
+	}
+
+	/**
+	 * Enable a membership source.
+	 * @param string $type
+	 * @param int $src_id
+	 */
+	public function enableAutomaticMembershipSource(string $type, int $src_id)
+	{
+		$assigned_by = ilStudyProgrammeAutoMembershipSource::SOURCE_MAPPING[$type];
+		$member_ids = $this->getMembersOfMembershipSource($type, $src_id);
+
+		foreach ($member_ids as $usr_id) {
+			if (!$this->getAssignmentsOfSingleProgramForUser($usr_id)) {
+				$this->assignUser($usr_id, $assigned_by);
+			}
+		}
+		$ams = $this->auto_memberships_repository->create($this->getId(), $type, $src_id, true);
+		$this->auto_memberships_repository->update($ams);
+	}
+
+	/**
+	 * Get member-ids of a certain source.
+	 * @param string $type
+	 * @param int $src_id
+	 * @throws InvalidArgumentException if $src_type is not in AutoMembershipSource-types
+	 * @return int[]
+	 */
+	protected function getMembersOfMembershipSource(string $src_type, int $src_id): array
+	{
+		$source_reader = $this->membersourcereader_factory->getReaderFor($src_type, $src_id);
+		return $source_reader->getMemberIds();
+	}
+
+
+	/**
+	 * Get all StudyProgrammes monitoring this membership-source.
+	 * @param int $cat_ref_id
+	 * @return ilObjStudyProgramme[]
+	 */
+	protected static function getProgrammesMonitoringMemberSource(string $src_type, int $src_id): array
```

Comment:
Die ganzen statischen methoden sind ja nicht so dolle. Gehören die vielleicht in ein neues objekt?

Reply:
Das ist im Wesentlichen dem EventHandling geschuldet...